superset/commands/database/update.py

```diff
@@ -75,127 +73,106 @@ def run(self) -> Model:  # pylint: disable=too-many-statements, too-many-branche
-                        except (
-                            SSHTunnelInvalidError,
-                            SSHTunnelUpdateFailedError,
-                            SSHTunnelDatabasePortError,
-                        ) as ex:
-                            # So we can show the original message
-                            raise ex
-                        except Exception as ex:
-                            raise DatabaseUpdateFailedError() from ex
-
-            # adding a new database we always want to force refresh schema list
-            # TODO Improve this simplistic implementation for catching DB conn fails
-            try:
-                schemas = database.get_all_schema_names(ssh_tunnel=ssh_tunnel)
-            except Exception as ex:
-                db.session.rollback()
-                raise DatabaseConnectionFailedError() from ex
-
-            # Update database schema permissions
-            new_schemas: list[str] = []
-
-            for schema in schemas:
-                old_view_menu_name = security_manager.get_schema_perm(
-                    old_database_name, schema
-                )
-                new_view_menu_name = security_manager.get_schema_perm(
-                    database.database_name, schema
-                )
-                schema_pvm = security_manager.find_permission_view_menu(
-                    "schema_access", old_view_menu_name
-                )
-                # Update the schema permission if the database name changed
-                if schema_pvm and old_database_name != database.database_name:
-                    schema_pvm.view_menu.name = new_view_menu_name
-
-                    self._propagate_schema_permissions(
-                        old_view_menu_name, new_view_menu_name
-                    )
-                else:
-                    new_schemas.append(schema)
-            for schema in new_schemas:
-                security_manager.add_permission_view_menu(
-                    "schema_access", security_manager.get_schema_perm(database, schema)
-                )
-
-            db.session.commit()
-
+            ssh_tunnel = self._handle_ssh_tunnel(database)
+            self._refresh_schemas(database, original_database_name, ssh_tunnel)
+        except SSHTunnelError as ex:
+            # allow exception to bubble for debugbing information
+            raise ex
         except (DAOUpdateFailedError, DAOCreateFailedError) as ex:
             raise DatabaseUpdateFailedError() from ex
+
         return database
 
-    @staticmethod
-    def _propagate_schema_permissions(
-        old_view_menu_name: str, new_view_menu_name: str
+    def _handle_ssh_tunnel(self, database: Database) -> SSHTunnel | None:
+        """
+        Delete, create, or update an SSH tunnel.
+        """
+        if "ssh_tunnel" not in self._properties:
+            return None
+
+        if not is_feature_enabled("SSH_TUNNELING"):
+            db.session.rollback()
```

Comment:
are there any hanging transactions here?

Reply:
There are, the database has been added before the SSH tunnel. I have a ticket to improve this and make it atomic.